### ISSUE TITLE
Kubelet CPU/Memory Usage linearly increases using CronJob

### ISSUE THREAD

**Issue by kaspernissen** [kubernetes/kubernetes#64137]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
For the past couple of months, we have been experiencing issues with kubelet on different nodes starts using more and more CPU and Memory. This happens somewhat randomly, but we think we have tracked it down to an issue with CronJob. This didn't happen before we started using CronJob in the beginning of the year. For the past two days, we had three separate occurrences of this issue.

**_Incident 1:_**
![screen shot 2018-05-22 at 09 40 17](https://user-images.githubusercontent.com/4429108/40348538-42e440de-5da4-11e8-884b-304ae6a7a049.png)
As seen in the graph above, the mode=system usage increases fairly linear over a period of 16-17 hours. When ssh'ing into the node and running top, it's apparent that the kubelet is using a lot of resources.

We try to dig deeper, by going through the kubelet logs at the time when the increase seems to begin (15:34), we find the following:
```
2018-05-20 15:34:08.000 | I0520 13:34:08.144400 21616 kubelet_pods.go:1080] Killing unwanted pod "bec-monitoring-1526823240-kvccs" |  
-- | -- | --
2018-05-20 15:34:08.000 | W0520 13:34:08.149174 21616 cni.go:265] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "edd71e53c8d95571cc8ff38cc852852a776739555ccf294d5b003015b326cb07" |  
2018-05-20 15:34:08.000 | weave-cni: unable to release IP address: 400 Bad Request: Delete: no addresses for edd71e53c8d95571cc8ff38cc852852a776739555ccf294d5b003015b326cb07 |  
2018-05-20 15:34:08.000 | I0520 13:34:08.188712 21616 qos_container_manager_linux.go:320] [ContainerManager]: Updated QoS cgroup configuration |  
2018-05-20 15:34:06.000 | I0520 13:34:06.176080 21616 kubelet_pods.go:1080] Killing unwanted pod "bec-monitoring-1526823240-kvccs" |  
2018-05-20 15:34:06.000 | E0520 13:34:06.234083 21616 cni.go:301] Error adding network: error setting up interface addresses: nsenter: cannot open /proc/20332/ns/net: No such file or directory |  
2018-05-20 15:34:06.000 | : exit status 1 |  
2018-05-20 15:34:06.000 | E0520 13:34:06.234106 21616 cni.go:250] Error while adding to cni network: error setting up interface addresses: nsenter: cannot open /proc/20332/ns/net: No such file or directory |  
2018-05-20 15:34:06.000 | : exit status 1 |  
2018-05-20 15:34:06.000 | weave-cni: error removing interface "eth0": nsenter: cannot open /proc/20332/ns/net: No such file or directory |  
2018-05-20 15:34:06.000 | : exit status 1 |  
2018-05-20 15:34:06.000 | E0520 13:34:06.332674 21616 remote_runtime.go:92] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "bec-monitoring-1526823240-kvccs_prod" network: error setting up interface addresses: nsenter: cannot open /proc/20332/ns/net: No such file or directory |  
2018-05-20 15:34:06.000 | : exit status 1 |  
2018-05-20 15:34:06.000 | E0520 13:34:06.332787 21616 kuberuntime_sandbox.go:54] CreatePodSandbox for pod "bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)" failed: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "bec-monitoring-1526823240-kvccs_prod" network: error setting up interface addresses: nsenter: cannot open /proc/20332/ns/net: No such file or directory |  
2018-05-20 15:34:06.000 | : exit status 1 |  
2018-05-20 15:34:06.000 | E0520 13:34:06.332819 21616 kuberuntime_manager.go:636] createPodSandbox for pod "bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)" failed: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "bec-monitoring-1526823240-kvccs_prod" network: error setting up interface addresses: nsenter: cannot open /proc/20332/ns/net: No such file or directory |  
2018-05-20 15:34:06.000 | : exit status 1 |  
2018-05-20 15:34:06.000 | E0520 13:34:06.332917 21616 pod_workers.go:182] Error syncing pod 75c1598e-5c32-11e8-9dae-0642ea7387e6 ("bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)"), skipping: failed to "CreatePodSandbox" for "bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)" with CreatePodSandboxError: "CreatePodSandbox for pod \"bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)\" failed: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod \"bec-monitoring-1526823240-kvccs_prod\" network: error setting up interface addresses: nsenter: cannot open /proc/20332/ns/net: No such file or directory\n: exit status 1" |  
2018-05-20 15:34:06.000 | I0520 13:34:06.335458 21616 qos_container_manager_linux.go:320] [ContainerManager]: Updated QoS cgroup configuration |  
2018-05-20 15:34:06.000 | I0520 13:34:06.871789 21616 kubelet.go:1871] SyncLoop (PLEG): "bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)", event: &pleg.PodLifecycleEvent{ID:"75c1598e-5c32-11e8-9dae-0642ea7387e6", Type:"ContainerDied", Data:"edd71e53c8d95571cc8ff38cc852852a776739555ccf294d5b003015b326cb07"} |  
2018-05-20 15:34:06.000 | W0520 13:34:06.871899 21616 pod_container_deletor.go:77] Container "edd71e53c8d95571cc8ff38cc852852a776739555ccf294d5b003015b326cb07" not found in pod's containers |  
2018-05-20 15:34:05.000 | W0520 13:34:05.008121 21616 prober.go:98] No ref for container "docker://98727dbc83cae5409b67c4c2c1c5494c7e4a2eb9bb413045e6e2c33cc445dc32" (default-http-backend-cbjnl_networking(3eb636ca-4935-11e8-9dae-0642ea7387e6):default-http-backend) |  
2018-05-20 15:34:05.000 | I0520 13:34:05.008194 21616 prober.go:101] Liveness probe for "default-http-backend-cbjnl_networking(3eb636ca-4935-11e8-9dae-0642ea7387e6):default-http-backend" errored: net/http: request canceled (Client.Timeout exceeded while reading body) |  
2018-05-20 15:34:05.000 | I0520 13:34:05.023101 21616 operation_generator.go:545] UnmountVolume.TearDown succeeded for volume "kubernetes.io/secret/#########" (OuterVolumeSpecName: "######") pod "75c1598e-5c32-11e8-9dae-0642ea7387e6" (UID: "75c1598e-5c32-11e8-9dae-0642ea7387e6"). InnerVolumeSpecName "#####". PluginName "kubernetes.io/secret", VolumeGidValue "" |  
2018-05-20 15:34:05.000 | I0520 13:34:05.097481 21616 reconciler.go:290] Volume detached for volume "######" (UniqueName: "kubernetes.io/secret/########") on node "ip-10-3-39-100.eu-west-1.compute.internal" DevicePath "" |  
2018-05-20 15:34:05.000 | I0520 13:34:05.850001 21616 kubelet.go:1871] SyncLoop (PLEG): "bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)", event: &pleg.PodLifecycleEvent{ID:"75c1598e-5c32-11e8-9dae-0642ea7387e6", Type:"ContainerDied", Data:"308625733a3b7c04115317413eb5ea7947e0a8082c3c014cf7751fc03c241267"} |  
2018-05-20 15:34:05.000 | W0520 13:34:05.850083 21616 pod_container_deletor.go:77] Container "308625733a3b7c04115317413eb5ea7947e0a8082c3c014cf7751fc03c241267" not found in pod's containers |  
2018-05-20 15:34:05.000 | I0520 13:34:05.853298 21616 kuberuntime_manager.go:392] No ready sandbox for pod "bec-monitoring-1526823240-kvccs_prod(75c1598e-5c32-11e8-9dae-0642ea7387e6)" can be found. Need to start a new one |  
2018-05-20 15:34:05.000 | W0520 13:34:05.855801 21616 cni.go:265] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "308625733a3b7c04115317413eb5ea7947e0a8082c3c014cf7751fc03c241267" |  
2018-05-20 15:34:05.000 | weave-cni: unable to release IP address: 400 Bad Request: Delete: no addresses for 308625733a3b7c04115317413eb5ea7947e0a8082c3c014cf7751fc03c241267

```

**_Incident 2_**
![screen shot 2018-05-22 at 09 52 35](https://user-images.githubusercontent.com/4429108/40349125-e3625964-5da5-11e8-97e6-e86a113288a4.png)
Same pattern as Incident 1. Linear increse in CPU and decrease in memory. 

Digging in to the logs at the starting point again: (multiple of the following lines) 
```

2018-05-20 13:47:12.000 | I0520 11:47:12.142453 21616 kubelet_pods.go:1080] Killing unwanted pod "bec-monitoring-1526816820-cftnn" |  
-- | -- | --
2018-05-20 13:47:12.000 | W0520 11:47:12.146932 21616 cni.go:265] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "28c448f8b5fffb62168c4bab2a779c200715874723a7c61fe2ffcf3c0cc78129" |  
2018-05-20 13:47:12.000 | weave-cni: unable to release IP address: 400 Bad Request: Delete: no addresses for 28c448f8b5fffb62168c4bab2a779c200715874723a7c61fe2ffcf3c0cc78129 |  
2018-05-20 13:47:12.000 | I0520 11:47:12.170373 21616 qos_container_manager_linux.go:320] [ContainerManager]: Updated QoS cgroup configuration |  
2018-05-20 13:47:10.000 | I0520 11:47:10.155868 21616 kubelet_pods.go:1080] Killing unwanted pod "bec-monitoring-1526816820-cftnn" |  
2018-05-20 13:47:10.000 | weave-cni: error removing interface "eth0": nsenter: cannot open /proc/7186/ns/net: No such file or directory |  
2018-05-20 13:47:10.000 | : exit status 1 |  
2018-05-20 13:47:10.000 | I0520 11:47:10.305582 21616 qos_container_manager_linux.go:320] [ContainerManager]: Updated QoS cgroup configuration |  
2018-05-20 13:47:10.000 | I0520 11:47:10.555789 21616 kubelet.go:1871] SyncLoop (PLEG): "bec-monitoring-1526816820-cftnn_prod(84d6a4fc-5c23-11e8-9dae-0642ea7387e6)", event: &pleg.PodLifecycleEvent{ID:"84d6a4fc-5c23-11e8-9dae-0642ea7387e6", Type:"ContainerDied", Data:"28c448f8b5fffb62168c4bab2a779c200715874723a7c61fe2ffcf3c0cc78129"} |  
2018-05-20 13:47:10.000 | W0520 11:47:10.555946 21616 pod_container_deletor.go:77] Container "28c448f8b5fffb62168c4bab2a779c200715874723a7c61fe2ffcf3c0cc78129" not found in pod's containers |  
2018-05-20 13:47:09.000 | I0520 11:47:09.535015 21616 kubelet.go:1871] SyncLoop (PLEG): "bec-monitoring-1526816820-cftnn_prod(84d6a4fc-5c23-11e8-9dae-0642ea7387e6)", event: &pleg.PodLifecycleEvent{ID:"84d6a4fc-5c23-11e8-9dae-0642ea7387e6", Type:"ContainerStarted", Data:"28c448f8b5fffb62168c4bab2a779c200715874723a7c61fe2ffcf3c0cc78129"} |  
2018-05-20 13:47:08.000 | I0520 11:47:08.143561 21616 kubelet_pods.go:1080] Killing unwanted pod "bec-monitoring-1526816820-cftnn" |  
2018-05-20 13:47:08.000 | W0520 11:47:08.148106 21616 cni.go:265] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "9e4eb4c2933fc27d59b63a5a9c23e273d7ba974a8cca516f1a4b12d0233dc54d" |  
2018-05-20 13:47:08.000 | weave-cni: unable to release IP address: 400 Bad Request: Delete: no addresses for 9e4eb4c2933fc27d59b63a5a9c23e273d7ba974a8cca516f1a4b12d0233dc54d
```

**_Incident 3_**
![screen shot 2018-05-22 at 09 58 14](https://user-images.githubusercontent.com/4429108/40349351-ac718708-5da6-11e8-9dca-ac9a2d8b7267.png)
Again the same pattern.

And the logs around that time is the same as before:
```
@timestamp | log |  
-- | -- | --
2018-05-21 14:08:06.000 | W0521 12:08:06.252658 3821 pod_container_deletor.go:77] Container "b775f90ece2d8674b6e033edd606fb119e66d55f8353f56c84d1badbb4e5812f" not found in pod's containers |  
2018-05-21 14:08:05.000 | I0521 12:08:05.134309 3821 kubelet_pods.go:1080] Killing unwanted pod "bec-monitoring-1526904480-cfchl" |  
2018-05-21 14:08:05.000 | weave-cni: error removing interface "eth0": nsenter: cannot open /proc/7009/ns/net: No such file or directory |  
2018-05-21 14:08:05.000 | : exit status 1 |  
2018-05-21 14:08:05.000 | I0521 12:08:05.212048 3821 kubelet.go:1871] SyncLoop (PLEG): "bec-monitoring-1526904480-cfchl_prod(9b836cca-5cef-11e8-9dae-0642ea7387e6)", event: &pleg.PodLifecycleEvent{ID:"9b836cca-5cef-11e8-9dae-0642ea7387e6", Type:"ContainerStarted", Data:"b775f90ece2d8674b6e033edd606fb119e66d55f8353f56c84d1badbb4e5812f"} |  
2018-05-21 14:08:05.000 | I0521 12:08:05.286498 3821 qos_container_manager_linux.go:320] [ContainerManager]: Updated QoS cgroup configuration |  
2018-05-21 14:08:04.000 | I0521 12:08:04.177863 3821 kubelet.go:1871] SyncLoop (PLEG): "bec-monitoring-1526904480-cfchl_prod(9b836cca-5cef-11e8-9dae-0642ea7387e6)", event: &pleg.PodLifecycleEvent{ID:"9b836cca-5cef-11e8-9dae-0642ea7387e6", Type:"ContainerDied", Data:"def4398c4393075f4fb36fcc4d55ae6bc9ca9439c7b0175d28facbf146a6ba61"} |  
2018-05-21 14:08:04.000 | W0521 12:08:04.177970 3821 pod_container_deletor.go:77] Container "def4398c4393075f4fb36fcc4d55ae6bc9ca9439c7b0175d28facbf146a6ba61" not found in pod's containers |  
2018-05-21 14:08:04.000 | I0521 12:08:04.181260 3821 kuberuntime_manager.go:392] No ready sandbox for pod "bec-monitoring-1526904480-cfchl_prod(9b836cca-5cef-11e8-9dae-0642ea7387e6)" can be found. Need to start a new one |  
2018-05-21 14:08:04.000 | W0521 12:08:04.183702 3821 cni.go:265] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "def4398c4393075f4fb36fcc4d55ae6bc9ca9439c7b0175d28facbf146a6ba61" |  
2018-05-21 14:08:04.000 | weave-cni: unable to release IP address: 400 Bad Request: Delete: no addresses for def4398c4393075f4fb36fcc4d55ae6bc9ca9439c7b0175d28facbf146a6ba61
```

We have a 3 cronjobs running at regurlar intervals, one of them is the mentioned container in the logs: bec-monitoring. We suspect that cronjobs may be the reason to this problem. 
`restartPolicy: Never` is set in all cronjobs.

The most aggressive cronjob is running every minute with the following configuration: (some details removed)
```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: bec-monitoring
spec:
  schedule: '*/1 * * * *'
  jobTemplate:
    spec:
      template:
        metadata:
          name: bec-monitoring
          labels:
            app: bec-monitoring
        spec:
          containers:
          - name: bec-monitoring
            image: <IMAGE>
            args:
            - ./bec
            - --identifiers
            - 12345678
          restartPolicy: Never
```

We have been looking into similar issues related to docker shim: https://github.com/kubernetes/kubernetes/issues/55620
https://github.com/kubernetes/kubernetes/pull/55641

But we don't see any id's in the docker shim folder, which is not referencing a currently running container. Except in one of the case where a cronjob container was exited  18 minutes prior - and therefore not relating to the beginning of the linear increase many hours before.

The solution to the increase, is to restart the kubelet, and everything seems to return to normal.

**What you expected to happen**:
The Kubelet not to use all of the resources. 

**How to reproduce it (as minimally and precisely as possible)**:
Unfortunately I haven't found a way to reproduce. 

**Anything else we need to know?**:
Running weavenet as CNI, with the following image: `weaveworks/weave-kube:2.2.0`

**Environment**:
- Kubernetes version (use `kubectl version`):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.10", GitCommit:"044cd262c40234014f01b40ed7b9d09adbafe9b1", GitTreeState:"clean", BuildDate:"2018-03-19T17:44:09Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
```
AWS
```
- OS (e.g. from /etc/os-release):
```
admin@ip-10-3-39-100:~$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
- Kernel (e.g. `uname -a`):
```
admin@ip-10-3-39-100:~$ uname -a
Linux ip-10-3-39-100 4.4.111-k8s #1 SMP Sun Jan 14 19:32:08 UTC 2018 x86_64 GNU/Linux
```
- Install tools:
```
kops
```

**Comment by zloo** [comment]
Hi @kaspernissen . I think we're experiencing something similar in our cluster (we're using 1.10.2 though).

Could you please look at `systemd-cgls -a` and look for lines that look like this:
`  ├─run-re016aa97d850410186c0dd748c267fdd.scope`
directly under `system.slice` ?

**Comment by zloo** [comment]
Hi everyone,

We believe we're having the same issue. We were able to identify that every run of a cronjob with a mounted secret causes the OS to leak one more cgroup, as evidenced by `systemd-cgls -a | egrep "run-.*scope"`

This in turn causes huge CPU and memory load for kubelet, at least in part because we scrape the cAdvisor statistics, which include cpu and memory usage for every cgroup present on the system. 

I was able to reproduce this issue with a simple CronJob with nodeSelector set for a specific node:
```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: test-cgroups-bug
  namespace: tt2-stable
spec:
  schedule: "*/1 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          nodeSelector:
            kubernetes.io/hostname: tt-k8s1-w69.ko.seznam.cz
          containers:
          - name: hello
            image: debian:stretch
            args:
            - /bin/sh
            - -c
            - date
            volumeMounts:
            - mountPath: /secret
              name: test-secret
          restartPolicy: OnFailure
          volumes:
          - name: test-secret
            secret:
              defaultMode: 420
              secretName: test-cgroup-secret
```

When I delete the secret from CronJob spec, the cgroup count stays the same.

We're using k8s 1.10.2, docker 17.3.1 and 17.3.2, we're running on our own infrastructure on Ubuntu 16.04, kernel 4.14. We use calico as our network plugin. 

Is there anything that we can do to help you debug the issue? This issue is extremely problematic for us because we need to drain and reboot nodes in order to fix the issue.

**Comment by dims** [comment]
cc @soltysh 

Do we want to mark this to block v1.11 release?

**Comment by soltysh** [comment]
> Do we want to mark this to block v1.11 release?

No, this is one of the reasons we need https://github.com/kubernetes/kubernetes/issues/17130 solved, so it's a known issue.

**Comment by zloo** [comment]
@soltysh I think that the dangling cgroups might be a separate issue, and #17130 and the cgroups issue both contribute to high cpu load and memory consumption.

**Comment by soltysh** [comment]
I agree with @zloo I spoke about the shared informers that I aware of.

**Comment by zloo** [comment]
I see that we have a PR for post-1.11 for the deepcopy issue, which is awesome!

Can we somehow test whether this also fixes the dangling cgroups issue? Was anyone able to reproduce it? Can we anyhow assist with something?

**Comment by zloo** [comment]
Alright, so we tested this further and it seems that a dangling cgroup is created whenever a pod with a volume referencing a secret is present in the pod spec. It doesn't even have to be mounted anywhere (i.e. "volumeMounts" is not present at all) 

Cronjobs only accentuate this problem because they create a lot of pods.

**Comment by zloo** [comment]
I have just confirmed that downgrading kernel to distribution supplied (ubuntu xenial) 4.4 kernel seems to solve the cgroup issue. We'll try to bisect what exactly causes this. The original reporter also seems to use a custom kernel, it might be related to a specific kernel option.

**Comment by zloo** [comment]
We tested these upstream ubuntu kernels:
linux-image-4.15.0-23-generic
linux-image-4.10.0-42-generic
linux-image-4.8.0-58-generic
And vanilla kernels with our kconfig:
4.4.138
4.4.0

All of the above caused the dangling cgroups to appear. The most likely culprit seems to be some incompatibility between systemd and kernel versions - it might be the issue that @CallMeFoxie mentioned.

As a hotfix, we will deploy a newer systemd version (232). As a long term fix, we will migrate to Bionic. Thanks to everyone for your help. If you want us to investigate something so you can write about it, or need anything else, feel free to contact me.

**Comment by bjhaid** [comment]
I can reproduce this outside k8s:

```
systemd-run --scope env | head -n1
Running as unit run-3685.scope.

ls /sys/fs/cgroup/memory/system.slice/run-3685.scope/
cgroup.clone_children  memory.failcnt         memory.max_usage_in_bytes        memory.oom_control          memory.stat            memory.use_hierarchy
cgroup.event_control   memory.force_empty     memory.move_charge_at_immigrate  memory.pressure_level       memory.swappiness      notify_on_release
cgroup.procs           memory.limit_in_bytes  memory.numa_stat                 memory.soft_limit_in_bytes  memory.usage_in_bytes  tasks
```

I believe this is because kubelet calls [`systemd-run`](https://github.com/kubernetes/kubernetes/blob/24ab69d358faeb3452bb76813fddd62afeefacb6/pkg/util/mount/nsenter_mount.go#L96-L110) when it mounts volumes, so this is likely a systemd-run bug

**Comment by bjhaid** [comment]
Can confirm does not exist on 232 systemd (stretch), so it probably go fixed somewhere in-between, but I can't find it in the release notes and I can confirm we don't have `systemd.unified_cgroup_hierarchy=1` configured as a boot param for either the older linux with systemd 217 or the newer one with 232 systemd

**Comment by rbojja66** [comment]
this affects us to, what's the timeline for resolving this issue.
we are facing this on our cluster running  systemd 238.

**Comment by bjhaid** [comment]
FWIW, below:

`find /sys/fs/cgroup/ -name run-*.scope -type d -exec rmdir {} \; `

is an hack that works

**Comment by nonken** [comment]
@bjhaid where would you execute this?

**Comment by bjhaid** [comment]
@nonken on the kubernetes nodes

**Comment by nonken** [comment]
@bjhaid Thank you! Much appreciated.

**Comment by dims** [comment]
long-term-issue (note to self)

**Comment by reaperes** [comment]
Same hacks for here. I used these systemd unit.
https://gist.github.com/reaperes/34ed7b07344ccc61b9570c46a3b4e564

**Comment by jayunit100** [comment]
So, to cross ref: https://github.com/kubernetes/kubernetes/issues/70324 , people seeing Kubelet CPU spiking + systemd / cgroup leakage might want to look at this or the other issue.  In our case kubelet CPU was going to 3000% within a few minutes and never returning from the part in the code where it started loading device plugins.  Im not sure which upstream issue this corresponds to , yet, though - because ultimately that 3000% CPU usage when straced was due to lots of lstats against `/sys/fs/.../burstable/...`.

**Comment by derekrprice** [comment]
@reaperes's systemd code seemed the cleanest and most surgical of all the workarounds that I've found documented for this and the related issues, so I've converted it into a DaemonSet that runs the fix hourly on every node in a cluster.  You could set any interval that you like, of course, but the script isn't very resource intensive and hourly seemed reasonable.  It actually takes about a day or so for the CPU loading to become noticeable in my cluster and a week or so for it to crash a node.
* `kubectl apply -f` https://github.com/derekrprice/k8s-hacks/blob/master/systemd-cgroup-gc.yaml

**Comment by edernucci** [comment]
@derekrprice I have three clusters on AKS running this issue for a year. I have used the trick from @bjhaid on one node today and can confirm that it works. I'll run your DaemonSet on the first cluster and post here the results asap.

**Comment by derekrprice** [comment]
@edernucci, I liked @reaperes's technique slightly better.  I called it "more surgical" because it checks that the pod is really gone before it clears a scope and it uses systemd to clear it, hopefully doing any other cleanup that systemd might need.  I won't pretend to understand everything that is going on, but @bjhaid's fix looks like it might be clearing scopes that could still be in use and neglecting any other GC that systemd might perform when clearing them itself.

Incidentally, my DaemonSet with his fix has been running in both my Staging and Production nodes for almost a week now and there is no evidence of the previous problem.  I have almost level CPU usage and no node crashes.

**Comment by Chengyudong** [comment]
I tested that this problem won't appear with docker 18 . I think it might be a bug with docker 17 not systemd.

**Comment by amagnus** [comment]
@derekrprice Thanks for the daemonset suggestion. We're running our clusters with the flag `--allow-privileged=false`, which prevents us from using `securityContext.privileged: true` in the spec. Do you see any alternative to access a host PID from the container?

**Comment by derekrprice** [comment]
Sorry, no.  The kernel/systemd bug that this works around requires cleanup
to run on the host node.  You could convert this to a cron job and install
it explicitly on each K8s host node but I was explicitly trying to avoid
that since we are in AKS and would have had to reinstall the job with every
node rebuild due to a K8s update and with every scale-up operation.

**Comment by goatmale** [comment]
We also seem to be having this problem. Thank you all for clearly documenting the cause and outlining workarounds.

**Comment by erik-wramner** [comment]
We also have this issue in AKS. Would be nice with a more permanent solution. Until then I guess we'll implement one of the workarounds - thanks for posting them!

**Comment by riking** [comment]
Looks like there's an agreement this is a systemd bug on Pod exit when there are Secret volume mounts. Someone claimed it was fixed on 232 but another user reported seeing it on 238.

Is there a bug reported upstream?

/area teardown
/remove-lifecycle stale

**Comment by derekrprice** [comment]
Just a quick update, we are still seeing a memory leak in AKS running K8s 1.14.6 on Ubuntu 16.04.6 (kernel 4.15.0-1060-azure).  After applying the work-around that I posted above, the memory leak becomes almost unnoticeable on a daily basis, but over 30 days, a much slower memory leak is clearly still present.  In fact, it took down our production cluster again today.  I wasn't running the most recent patches to my work-around DaemonSet as suggested by @goatmale, but I don't think that those patches are related to the remaining problem.  FYI, memory usage is still increasing by about 1-2% per day in a 6 node, 24GB cluster.  The cluster went down at about 60% overall usage.

I've implemented a second script to perform weekly reboots:
- `kubectl apply -f` https://github.com/derekrprice/k8s-hacks/blob/master/regular-reboot.yaml

**Comment by zloo** [comment]
Hi @derekrprice - I'm pretty sure you're hitting the same issue as I did a year ago. Look at my comment: https://github.com/kubernetes/kubernetes/issues/64137#issuecomment-398784448

Upgrading to a newer systemd and/or upgrading to new Ubuntu fixes the issue. It's a kernel(cgroup implementation) and systemd incompatibility.

**Comment by derekrprice** [comment]
@zloo I'm in AKS, so my kernel version is supposed to be managed and their support dragging their feet on my requests to update systemd.  Your comment lists incompatible versions, but what kernel and systemd versions did you settle on?  Maybe if I go in armed with that data it will help with my support ticket.  Thanks!

**Comment by CallMeFoxie** [comment]
@derekrprice Most info I had (and that we settled on with @zloo ) is in this post https://github.com/kubernetes/kubernetes/issues/64137#issuecomment-398715410 if that helps?

**Comment by derekrprice** [comment]
@CallMeFoxie Ah, yes, sorry.  I read @zloo's post too fast.  Sounds like if I can get to systemd 232 it might be fixed.  Thanks!

Incidentally, [Debian #912411](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=912411) may also be the same bug.  That one's reported against systemd 230.

**Comment by CallMeFoxie** [comment]
yup, we "fought" in the same company with @zloo :) they were using Ubuntu Xenial with newer custom kernel and we were using Debian Stretch with stock kernel.
Looking at that Debian bug it looks like backports so yeah, some odd combination of kernel and systemd :).

**Comment by erik-wramner** [comment]
@derekrprice we are also in AKS and are doing reboots with `drain`, `vm restart` and `uncordon`. It is messy. In your script you are just updating `/host/var/run/reboot-required`. Is that enough to make AKS restart the node with drain and uncordon automatically? Sounds too good to be true.

Also if you managed to convince support to do something about it I would appreciate a heads up as that could help our discussions with them.

**Comment by derekrprice** [comment]
@erik-wramner, it's not too good to be true if you've installed [kured](https://github.com/weaveworks/kured).  Once that is installed, just touch that file and you will get a controlled rolling reboot.  This is the method that MS recommends for [installing kernel and security patches](https://docs.microsoft.com/en-us/azure/aks/node-updates-kured).

No luck with MS support yet.  As near as I can tell, nothing is likely to happen until the next major kernel upgrade, and who knows when that will happen.

**Comment by edernucci** [comment]
AKS friends, please see if this is not your real issue (it was mine): https://github.com/Azure/AKS/issues/1373

**Comment by barney-s** [comment]
I assume this is applicable to Pods created by other controller as well. Could we change the title to reflect what we know.

**Comment by mfoliveira** [comment]
On Ubuntu 16.04 (Xenial) with the 4.15 (HWE) kernel this issue should have been fixed on systemd packages version `229-4ubuntu21.23` (LP bugs #[1847512](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1847512) and #[1846787](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1846787)) with this systemd [commit](https://github.com/systemd/systemd/commit/d8fdc62037b5) `d8fdc62037b5 (core: use an AF_UNIX/SOCK_DGRAM socket for cgroup agent notification)`.

**Comment by riking** [comment]
Looks like the systemd fix stuck.
